src/common/ad_mem.c

```diff
@@ -1626,19 +1834,160 @@ arena_reserve_addr(struct ad_arena *arena, daos_size_t size, struct ad_reserv_ac
+	struct ad_group_df	*tmp = NULL;
+	int			 start;
+	int			 end;
+	int			 cur;
+	bool			 found = false;
+	int			 weight;
+	int			 rc;
+
+	/* remove this group from addr sort groups */
+	for (start = cur = 0, end = arena->ar_grp_nr - 1; start <= end; ) {
+		cur = (start + end) / 2;
+		gd = arena->ar_addr_sorter[cur];
+
+		if (gd->gd_addr == group->gp_df->gd_addr) {
+			found = true;
+			break;
+		}
+
+		if (gd->gd_addr < group->gp_df->gd_addr) {
+			rc = -1;
+		} else {
+			rc = 1;
+		}
+
+		if (rc < 0)
+			start = cur + 1;
+		else
+			end = cur - 1;
+	}
+
+	D_ASSERT(found);
+	if (cur != arena->ar_grp_nr - 1)
+		memmove(&arena->ar_addr_sorter[cur], &arena->ar_addr_sorter[cur + 1],
+			(arena->ar_grp_nr - cur - 1) * sizeof(arena->ar_addr_sorter[0]));
+
+	weight = group_weight(group->gp_df);
+	found = false;
+	/* remove this group from size sort groups */
+	gd = group->gp_df;
+	for (start = cur = 0, end = arena->ar_grp_nr - 1; start <= end; ) {
+		cur = (start + end) / 2;
+		tmp = arena->ar_size_sorter[cur];
+		/* the same unit size */
+		if (tmp->gd_unit == gd->gd_unit) {
+			int	w;
+
+			if (tmp == gd) {
+				found = true;
+				break;
+			}
+
+			w = group_weight(tmp);
+			if (w < weight)
+				rc = -1;
+			else if (w > weight)
+				rc = 1;
+			else /* group address */
+				rc = tmp->gd_addr < gd->gd_addr ? -1 : 1;
+
+		} else if (tmp->gd_unit < gd->gd_unit) {
+			rc = -1;
+		} else {
+			rc = 1;
+		}
+
+		if (rc < 0)
+			start = cur + 1;
+		else
+			end = cur - 1;
+	}
+	D_ASSERT(found);
+	if (cur != arena->ar_grp_nr - 1)
+		memmove(&arena->ar_size_sorter[cur], &arena->ar_size_sorter[cur + 1],
+			(arena->ar_grp_nr - cur - 1) * sizeof(arena->ar_size_sorter[0]));
+
+	arena->ar_grp_nr--;
+}
+
+static inline int gp_df2index(struct ad_group *group)
+{
+	struct ad_arena_df *ad = group->gp_arena->ar_df;
+
+	return group->gp_df - &ad->ad_groups[0];
+}
+
+static int arena_tx_reset_group(struct ad_tx *tx, struct ad_group *group)
+{
+	struct ad_arena		*arena = group->gp_arena;
+	struct ad_arena_df	*ad = arena->ar_df;
+	struct ad_group_df	*gd = group->gp_df;
+	int			 rc;
+	struct ad_free_act	*act;
+
+	/* not published yet */
+	if (group->gp_unpub && !group->gp_publishing)
+		return 0;
+	if (gd->gd_unit_free != gd->gd_unit_nr)
+		return 0;
+
+	D_ALLOC_PTR(act);
+	if (!act)
+		return -DER_NOMEM;
+
+	/* lock the bit and prevent it from being reused before commit */
+	setbits64(arena->ar_bmap_rsv, group->gp_bit_at, group->gp_bit_nr);
+	setbits64(arena->ar_bmap_rsv1, gp_df2index(group), 1);
+	D_DEBUG(DB_TRACE, "resetting group=%p, bit_at=%d, bits_nr=%d\n",
+		group, group->gp_bit_at, group->gp_bit_nr);
+
+	rc = ad_tx_clrbits(tx, ad->ad_bmap, group->gp_bit_at, group->gp_bit_nr);
+	if (rc)
+		goto failed;
+
+	group->gp_reset = 1;
+	arena_remove_sort_group(group);
+	gd->gd_addr = 0;
+	rc = ad_tx_set(tx, gd, 0, sizeof(*gd), AD_TX_REDO | AD_TX_LOG_ONLY);
+	if (rc)
+		goto failed;
+
+	group_addref(group);
+	act->fa_group = group;
+	act->fa_at    = group->gp_bit_at;
```

Comment:
if this is unused, we probably should not set it at all.

Reply:
OK